### PR TITLE
Add yarn

### PR DESCRIPTION
**Pull request by jubick1337**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/nlp/language_modeling/prepare_truncated_data.py [hunk 1]
```diff
@@ -51,34 +60,44 @@ def _process_line(line, task, tokenizer, max_seq_length, tokens_to_generate_extr
...
         raise ValueError(f"{data_dir} or {task_file} not found")
-    
+
     tokens_to_generate = TASKS[task]["tokens_to_generate"]
     prompt = None if prompt in ["None", None, ""] else prompt
     if prompt is not None:
         # when passing via command line, "\n" is changed to "\n\n"
         prompt = prompt.replace("\\n", "\n")
         assert "{prompt}" in prompt, "prompt must contain {prompt} for replacement"
         tokens_to_generate += len(tokenizer.text_to_tokens(prompt.replace("{prompt}", "")))
```
Comment (hsiehjackson) [inline examples/nlp/language_modeling/prepare_truncated_data.py:81]: What does this mean? Do we have a example how we get `tokens_to_generate` if we have a prompt_template?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 2]
```diff
@@ -34,11 +75,57 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
...
+                1 - linear_ramp_mask(low, high, self.dim // 2).float()
+            ) * self.extrapolation_factor  # Get n-d rotational scaling corrected for extrapolation
+            inv_freq = inv_freq_interpolation * (1 - inv_freq_mask) + inv_freq_extrapolation * inv_freq_mask
+
+            self.register_buffer("inv_freq", inv_freq)
+            self.mscale = float(
+                get_mscale(self.seq_len_interpolation_factor) * self.attn_factor
+            )  # Get n-d magnitude scaling corrected for interpolation
+
+            self.max_seq_len_cached = max_position_embeddings
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:106]: If we already have `seq_len_interpolation_factor`, do we still need `max_position_embeddings`?

File: nemo/collections/nlp/modules/common/megatron/language_model.py [hunk 3]
```diff
@@ -563,7 +587,16 @@ def __init__(
             if rotary_percentage < 1:
                 rotary_dim = int(rotary_dim * rotary_percentage)
             self.rotary_pos_emb = RotaryEmbedding(
-                rotary_dim, seq_len_interpolation_factor=seq_len_interpolation_factor
+                rotary_dim,
+                seq_len_interpolation_factor=seq_len_interpolation_factor,
+                # base=base,
+                # pretrain_max_positional_embeddings=pretrain_max_positional_embeddings,
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/language_model.py:593]: we can pass `max_position_embeddings` from line 546 as `pretrain_max_positional_embeddings`.

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 4]
```diff
@@ -34,11 +75,57 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
...
+            emb = torch.cat((freqs, freqs), dim=-1)
+            self.register_buffer('emb', emb)
+
+        else:
+            inv_freq = 1.0 / (self.base ** (torch.arange(0, dim, 2).float() / dim))
+            self.register_buffer('inv_freq', inv_freq)
 
     def forward(self, max_seq_len, offset=0):
+        if self.use_yarn:
+            if max_seq_len > self.max_seq_len_cached:
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:119]: I think this would not happen if we do continue training. The author did this for inference length longer than 1) pre-trained sequence length for pre-trained model or 2) continue training sequence length for fine-tuned model. And they dynamically change the scale based on the input `max_seq_len` from [here](https://github.com/jquesnelle/yarn/blob/master/scaled_rope/LlamaDynamicYaRNScaledRotaryEmbedding.py#L67).   

If we want to evaluate > 4K on our pre-trained 4K length models, we may need dynamic yarn. 
If we want to evaluate 8K or 16K on continue training 8K or 16K length models, we don't need dynamic yarn.

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 5]
```diff
@@ -34,11 +75,57 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None):
...
+
+                t = torch.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype).to(self.inv_freq.device)
+                freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
+                # Different from paper, but it uses a different permutation in order to obtain the same calculation
+                emb = torch.cat((freqs, freqs), dim=-1).to(self.inv_freq.device)
+                self.register_buffer('emb', emb)
+            return rearrange(self.emb, 'n d -> n 1 1 d')
+
         seq = torch.arange(max_seq_len, device=self.inv_freq.device) + offset
         if self.seq_len_interpolation_factor is not None:
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:130]: We can simply add `if self.seq_len_interpolation_factor is not None and not self.use_yarn:`

File: examples/nlp/language_modeling/prepare_truncated_data.py [hunk 6]
```diff
@@ -24,24 +24,33 @@
...
     context_len = len(context_tokens)
     truncation_seperator = line['truncation_seperator']
     truncation_seperator_tokens_len = len(tokenizer.text_to_tokens(truncation_seperator))
-    
+
     total_len = context_len + tokens_to_generate_extra + question_len
     if total_len > max_seq_length:
-        truncated_text = tokenizer.tokens_to_text(context_tokens[:(max_seq_length - tokens_to_generate_extra - truncation_seperator_tokens_len - question_len)]) + truncation_seperator + question
+        truncated_text = (
+            tokenizer.tokens_to_text(
```
[the remark this answers is not in the document]
Comment (jubick1337) [reply examples/nlp/language_modeling/prepare_truncated_data.py:46]: @ekmb

File: examples/nlp/language_modeling/prepare_truncated_data.py (hunk 1 above)
Comment (jubick1337) [reply examples/nlp/language_modeling/prepare_truncated_data.py:81]: @ekmb

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py (hunk 2 above)
Comment (jubick1337) [reply nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:106]: probably not, I'll remove it and replace usages of `max_position_embeddings` with `seq_len_interpolation_factor * pretrain_max_positional_embeddings`

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 7]
```diff
@@ -112,10 +122,10 @@ def forward(self, max_seq_len, offset=0):
                 t = torch.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype).to(self.inv_freq.device)
                 freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:123]: Don't we need to dynamically change scale to `max_seq_len / self. max_position_embeddings`?

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 8]
```diff
@@ -35,15 +79,68 @@ def __init__(self, dim: int, seq_len_interpolation_factor: int = None, pretraine
...
+
     def forward(self, max_seq_len, offset=0):
+        if self.use_yarn:
+            if max_seq_len > self.max_seq_len_cached:
+                self.max_seq_len_cached = max_seq_len
+                self.yarn(max_seq_len / self.max_positional_embeddings)
+                t = torch.arange(self.max_seq_len_cached, dtype=self.inv_freq.dtype).to(self.inv_freq.device)
+                freqs = torch.einsum("i , j -> i j", t, self.inv_freq)
+                # Different from paper, but it uses a different permutation in order to obtain the same calculation
+                emb = torch.cat((freqs, freqs), dim=-1).to(self.inv_freq.device) * self.mscale
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:137]: Do we need to remove self.mscale? although this doesn't affect continue training

File: nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py [hunk 9]
```diff
@@ -81,5 +178,8 @@ def apply_rotary_pos_emb(t, freqs):
     t, t_pass = t[..., :rot_dim], t[..., rot_dim:]
     # first part is cosine component
     # second part is sine component, need to change signs with _rotate_half method
-    t = (t * freqs.cos()) + (_rotate_half(t) * freqs.sin())
+    if mscale:
+        t = (t * freqs.cos() * mscale) + (_rotate_half(t) * freqs.sin() * mscale)
+    else:
+        t = (t * freqs.cos()) + (_rotate_half(t) * freqs.sin())
     return torch.cat((t, t_pass), dim=-1)
```
Comment (hsiehjackson) [inline nemo/collections/nlp/modules/common/megatron/position_embedding/rotary_position_embedding.py:185]: I think maybe we should multiply mscale to all `torch.cat((t, t_pass), dim=-1)` since the original yarn has rope for all dimension but we only have 50% dimension in our models.

**Review by hsiehjackson — CHANGES_REQUESTED** [review]
Some minor changes. The implementation of yarn looks good to me.

**Review by hsiehjackson — APPROVED** [review]
LGTM!
